[M]1712.cpp

```diff
@@ -0,0 +1,84 @@
+//Problem: 1712. Ways to Split Array Into Three Subarrays - Medium
+//Link: https://leetcode.com/problems/ways-to-split-array-into-three-subarrays/
+
+/*
+1. Nhận xét:
+- Nếu sử dụng mảng prefix sum, có thể tính tổng left, mid, right với độ phức tạp O(1)
+- Khi dãy left càng tăng, thì số lựa chọn cho dãy mid và dãy right càng giảm.
+- Khi ta chốt dãy left cố định, thì ta có thể sử dụng một mốc để chia 2 dãy mid và right. Mốc này sẽ nằm trong 1 khoảng LIÊN TỤC. 
+Mục tiêu của ta là phải đi tìm biên trái và biên phải của “khoảng mốc” này.
+2. Phương án:
+- Thử tất cả các vị trí có thể cho dãy left.
+- Với mỗi vị trí này, ta sử dụng binary search để tìm biên trái và biên phải của “khoảng mốc”:
+	+ Với biên trái: điều kiện tất cả những điểm ở bên phải của biên có total mid > total left
+	+ Với biên phải: điều kiện tất cả những điểm ở bên trái của biên có total right > total mid
+- Nếu đến một vị trí của left mà không thể tìm được biên phải và biên trái phù hợp 
+=> ta có thể kết luận ngay, vì các vị trí left tiếp theo cũng đều không thỏa mãn
+3. Thơi gian code và debug: ~ 2 tiếng
+*/
+
+class Solution {
+public:
+    int n;
+    vector<int> pref;
+    const int mod = 1e9 + 7;
+    
+    int findLeft(int l, int r, int low) {
```

Comment:
Không nên khai báo biến l, r và sửa l, r ở bên trong hàm. Đấy ko phải cách code tốt.

Reply:
Cái này anh chưa hiểu lắm nhé, vì anh thấy nhiều người khác vẫn làm thế. Trường hợp không sửa giá trị thì họ sẽ khai báo const &. Còn khi để thế này thì đôi khi họ vẫn sửa. Nếu không thì chú bảo sửa code của anh kiểu gì?

Reply:
ý là a sẽ  code như thế này: 
int findLeft(int startIndex, low) {
   int l = startIndex, r = n - 1;
   while(r - l > 1) {...}
}
Chú sẽ ko phải gán lại giá trị của l, r như ở 2 dòng 68 với 75 nữa (logic bị chia chẻ ở nhiều nơi, code thế này đóng gói hơn).

Reply:
Các ngôn ngữ hướng đối tượng sẽ báo lỗi lint nếu như chú khai báo 1 biến ở hàm và trong hàm lại gán lại giá trị của biến đấy.

Reply:
> ý là a sẽ code như thế này:
> int findLeft(int startIndex, low) {
> int l = startIndex, r = n - 1;
> while(r - l > 1) {...}
> }
> Chú sẽ ko phải gán lại giá trị của l, r như ở 2 dòng 68 với 75 nữa (logic bị chia chẻ ở nhiều nơi, code thế này đóng gói hơn).

Cái này thì anh phải đính chính một chút nhé:
- Ở dòng 68, với 75 là 2 đoạn anh đổi cái biên. Ở hàm tìm biên trái anh sẽ return l, trong khi hàm tìm biên phải anh sẽ return r => Anh phải đổi logic biên nên mới chỉnh lại. Chỗ này anh thấy cũng hơi lủng củng, nhờ chú góp ý cho anh?
- Còn ý thứ 2 thì anh phải confirm lại, với C++ thì anh đang truyền tham số theo kiểu Pass by value, tức là khi anh truyền l và r vào thì thực chất trong hàm đó nó sẽ tạo 2 biến l và r riêng biệt, độc lập với bên ngoài. Khi sửa giá trị l và r trong hàm này thì bên ngoài không bị ảnh hưởng gì cả => Nếu làm như chú:
> int findLeft(int startIndex, low) {
> int l = startIndex, r = n - 1;
> while(r - l > 1) {...}
thì cái biến startIndex sẽ được tạo mới trong hàm, xong sau đó hoàn toàn không sử dụng đến, như vậy anh cảm thấy hơi thừa. Chú nghĩ sao?

Reply:
1. Tốt nhất đừng dùng lại biến giữa các hàm như vậy. Có thể viết 1 hàm chung chung kiểu như sau:
findLeft(int startIndex, int endIndex, int low) {
    int l = startIndex, r = endIndex;
    //do some logic here
}
Và bên dưới ở dòng 70 chỉ cần pass int left = findLeft(i, r - 2, prefix[i]); right = findRight(i + 1, n - 1, prefix[i]); Code nó sẽ clean hơn.
2. Đồng ý là C++ nó pass by value, chú sửa nó sẽ ko ảnh hưởng nhưng vẫn ko nên làm thế. Nó gây confuse giữa biến local và biến global.

Reply:
Ok. Got it!